src/hal/tiva_c/spi.rs

```diff
@@ -0,0 +1,304 @@
+  pub peripheral: SpiId,
+
+  /// Bus frequency in Hz. Peripheral clock scaling is calculated from this value to a possibly non-exact match
+  pub frequency: u32,
+}
+
+/**
+Structure describing a single SPI interface
+
+# Examples
+
+Create an SPI interface on SSI0 with a run frequency of 4MHz
+
+```
+#![crate_type = "staticlib"]
+#![feature(plugin, start, core_intrinsics)]
+#![no_std]
+#![plugin(macro_platformtree)]
+
+extern crate zinc;
+
+use zinc::hal::spi::Spi;
+use zinc::hal::tiva_c::spi;
+
+platformtree!(
+  tiva_c@mcu {
+    // Tiva C ends up with an 80MHz clock from 16MHz external xtal and x5 PLL
+    clock {
+      source = "MOSC";
+      xtal   = "X16_0MHz";
+      pll    = true;
+      div    = 5;
+    }
+
+    gpio {
+      a {
+        spi_ck@2 {
+          direction = "out";
+          function  = 2;
+        }
+
+        spi_tx@5 {
+          direction = "out";
+          function  = 2;
+        }
+      }
+    }
+  }
+
+  os {
+    single_task {
+      loop = "run";
+    }
+  }
+);
+
+fn run() {
+  let spi = spi::Spi::new(spi::SpiConf {
+    peripheral: spi::SpiId::Spi0,
+    frequency: 4_000_000
+  });
+
+  loop {
+    spi.write('a' as u8);
+  }
+}
+```
+*/
+#[derive(Clone, Copy)]
+pub struct Spi {
+  /// SSI registers
+  regs: &'static reg::Ssi,
+}
+
+impl Spi {
+  /// Create and setup an SPI interface.
+  pub fn new(config: SpiConf) -> Spi {
+
+    let (periph, regs) = match config.peripheral {
+      SpiId::Spi0 => (sysctl::periph::ssi::SSI_0, reg::SSI_0),
+      SpiId::Spi1 => (sysctl::periph::ssi::SSI_1, reg::SSI_1),
+      SpiId::Spi2 => (sysctl::periph::ssi::SSI_2, reg::SSI_2),
+      SpiId::Spi3 => (sysctl::periph::ssi::SSI_3, reg::SSI_3),
+    };
+
+    let spi = Spi { regs: get_reg_ref(regs) };
+
+    // Make sure peripheral clock gating is enabled
+    periph.ensure_enabled();
+
+    spi.configure(config);
+
+    spi
+  }
+
+  /// Configure the SSI into SPI mode. Currently hard coded at 3.2MHz for 16MHz PIOSC
+  fn configure(&self, config: SpiConf) {
+    // Disable peripheral so we can configure it
+    self.regs.ssicr1.set_sse(false);
+
+    self.regs.ssicr1
+      .set_sse(false)
+      .set_ms(false)
+      .set_lbm(false);
+
+    // Set clock rate
+    self.set_frequency(config.frequency);
+
+    self.regs.ssicr0
+      .set_sph(false)
+      .set_spo(false)
+      .set_frf(0)     // Put SSI into SPI mode
+      .set_dss(0x7);  // 8 bit frames
+
+    // Turn on peripheral now that we've configured it
+    self.regs.ssicr1.set_sse(true);
+  }
+
+  /// Set SPI frequency
+  ///
+  /// This function computes the divisor and exponent (for lack of a better name). These are stored in `ssicpsr.cpsdvsr` and `ssicr0.scr` respectively.
+  /// The clock rate formula (taken from the datasheet is) `ClockRate = SysClk / (CPSDVSR * (1 + SCR))` where SysClk is the system clock in Hz.
+  fn set_frequency(&self, freq: u32) {
+    let sysclk = sysctl::clock::sysclk_get() as u16;
```

Comment:
I'm still quite new to Rust (and also quite tired) so I didn't twig that `usize` should cast to `u32`. This and some other stuff is fixed/changed in commits I've just pushed.